### ISSUE TITLE
How are NavigationView icons implemented?

### ISSUE THREAD

**Issue by rajendrasp** [microsoft/WinUI-Gallery#693]
Hello,

I am trying to understand how icons are displayed in left NavigationView on this app and could not find any reference in the source. I saw that in the ControlInfoData.json file image path like this - "ImagePath": "". What is this? And there is a NavigationViewItemIcons folder but all icons on app look different than these pngs. Also, are these really pngs? Basically I want to understand how I can insert icons for my app which I am modelling on this one. I do not know if this is the right place to put this question. Sorry about that. Any help will be appreciated.

Thanks.

**Comment by chingucoding** [comment]
Hi @rajendrasp, in the XAML Controls Gallery the NavigationView icons are implemented using the ControlInfoData.json file. the "ImagePath" property is used for this. If the ImagePath is not a path but rather a special character, we will just use that. The thing is that those special characters are actually Segoe MDL2 icons. This is also the reason why they are being displayed as a box in Visual Studio. If the path is being specified, we just use the image specified.

You can find the code on how we handle the logic and create the actual NavigationViewItems here: https://github.com/microsoft/Xaml-Controls-Gallery/blob/c136d323ddefa0f8cfab125abdb66c350fcb913f/XamlControlsGallery/Navigation/NavigationRootPage.xaml.cs#L161

Hope this helps you, if you have any other questions, feel free to ask.

**Comment by rajendrasp** [comment]
Thank you for your reply @chingucoding. I see it. But how can I insert that special character icon in json file? If I want to use one of Segoe MDL2 icons? Also if I replaced that special character with a valid path, application is crashing. Why can I not specify a different png as an icon there? I want to mainly use a custom one and not one of Segoe MDL2 icons.

**Comment by chingucoding** [comment]
The app is probably crashing because either the path is not valid with regard to the apps installation folder or the app doesn't have access to it. The image is being resolved here so it might be good to see if your image matches all the criteria (e.g. is it a PNG?).
https://github.com/microsoft/Xaml-Controls-Gallery/blob/c136d323ddefa0f8cfab125abdb66c350fcb913f/XamlControlsGallery/Navigation/NavigationRootPage.xaml.cs#L222-L231

Regarding the special character, you want to insert the unicode character for the icon you want to use. To learn more about this, I think this page summarizes it good: https://www.fileformat.info/tip/microsoft/enter_unicode.htm
Note that the icon you want to use also has to be available in the font that you are using to display the character.

**Comment by rajendrasp** [comment]
I could get the unicode character to work but it still crashes randomly when I used this character from range PUA EA00-EC00. I am using the character EAFC which is a market icon. 
And I could not get my png to work. My icon file is png - I am attaching it here. And I am giving the path as "ImagePath": "ms-appx:///Assets/market.png",
![market2](https://user-images.githubusercontent.com/51284446/117569594-094ceb00-b0e4-11eb-9ec9-efdfddc80021.png)

**Comment by chingucoding** [comment]
> I could get the unicode character to work but it still crashes randomly when I used this character from range PUA EA00-EC00. I am using the character EAFC which is a market icon.

That's strange. I used this unicode character and it worked fine for me:  

> And I could not get my png to work. My icon file is png - I am attaching it here. And I am giving the path as "ImagePath": "ms-appx:///Assets/market.png",

Did you also add a reference to the icon in your project? If you don't add a reference to it, the icon will not be copied to the output folder and as such, the app will crash as it can't find the icon.

**Comment by rajendrasp** [comment]
> That's strange. I used this unicode character and it worked fine for me: 
Yes I used same unicode character and it is crashing on alternate launch of app. One successful launch and then next crash.

> Did you also add a reference to the icon in your project? If you don't add a reference to it, the icon will not be copied to the output folder and as such, the app will crash as it can't find the icon.

Yes, I missed this step. Now I am able to use the icon. But again same crash on alternate launch.
Crash happens for me for both unicode as well as BitmapIcon. There seems to be a pattern. Crash happens when the icon is rendered. If I make sure that icon is not rendered on launch, then it does not crash. Do I need to release the resource somehow? Thanks for your help.

**Comment by rajendrasp** [comment]
I think I am using UWP-WinUI2 one since I did not install the project reunion extension. But how can I confirm which one I am using? Also, is there a difference in the UI that is provided by the two? I think both of them give us the Windows 10 UI right? Also I know that Project reunion gives us fluent design. Does WinUI 2 not give us fluent design?

And should I try WinUI3 to see if it solves the problem? Also, I can share the repo with you with my code if you are okay with it.

**Comment by rajendrasp** [comment]
I am not using Windows 10 Insider Build. Following is the branch (iconCrash) with my code. I have changed the project and solution name.
https://github.com/rajendrasp/Xaml-Controls-Gallery/tree/iconCrash

**Comment by chingucoding** [comment]
I just ran this on my machine, restarted the app multiple times with and without debugger, it always starts just fine. One thing I noticed is that you set the image path to neither a unicode character or an image (link: https://github.com/rajendrasp/Xaml-Controls-Gallery/commit/86d270764636b70d2449570d92c001cd25cca84d#diff-79400188af461db868bd481690273a1b5ee3b59cd92313d1a5cdb2e7d6edcf06R327). While the app is not crashing on my machine at all, maybe this is causing the crash on your side?

These are just guesses but maybe some of these things help:
- Are you accidently starting x86 on a x64 PC (or similar)?
- Use release instead of debug or vice versa
- Fresh clone your repository and try that? Maybe there is some form of locked file that is causing the issue? At least I had this at time where I wasn't able to build or start an app because of some oddities regarding files being locked.
